src/rmkit/ui/events.cpy

```diff
@@ -0,0 +1,185 @@
+#include <cmath>
+
+#include "../input/events.h"
+#include "../util/signals.h"
+#include "timer.h"
+
+namespace ui:
+  PLS_DEFINE_SIGNAL(MOUSE_EVENT, input::SynMotionEvent)
+  class MOUSE_EVENTS:
+    public:
+    MOUSE_EVENT enter
+    MOUSE_EVENT leave
+    MOUSE_EVENT down
+    MOUSE_EVENT up
+    MOUSE_EVENT click
+    MOUSE_EVENT hover
+    MOUSE_EVENT move
+  ;
+
+  PLS_DEFINE_SIGNAL(KEY_EVENT, input::SynKeyEvent)
+  class KEY_EVENTS:
+    public:
+    KEY_EVENT pressed
+  ;
+
+  class GESTURE_EVENTS:
+    public:
+    MOUSE_EVENT long_press
+    MOUSE_EVENT single_click
+    MOUSE_EVENT double_click
+    MOUSE_EVENT drag_start
+    MOUSE_EVENT dragging
+    MOUSE_EVENT drag_end
+
+    // The threshold area that determines if this is a touch-based gesture
+    // (e.g. long_press, double_click) or a motion-based gesture (e.g. drag).
+    int touch_slop_size = 50
```

Comment:
I thought a lot about separating touch events (long_press, double_click, single_click) from the dragging events, since they seem like separate concepts. But I looked around at how some other toolkits deal with gestures, and it seemed like there was some momentum behind (a) lumping all the gestures together, and (b) having some kind of threshold where touch events are within the threshold, and motion events (drag, fling, etc.) are outside the threshold.

Reply:
it seems fine to put them together

what does `touch_slop_size` stand for?

Reply:
What I'm going for is something like this: the "ideal" touch is down and up at the same position, but in real life that doesn't actually happen, so "slop" is an extra allowance for real life touches.

I could probably call these two `touch_threshold` and `dragging_threshold` if that sounds better to you.

Reply:
i just wasn't sure if it was a typo (slope?), whichever name is fine. maybe 'allowance'? i'm not sure what word would be used here normally.

Reply:
how about touch_threshold? I also added a bit more detail to this comment to hopefully make it clearer what the value means